### PR TITLE
support base job worker framework

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut 

I will add test later.

### CODE REVIEW

**Comment by siddontang** [comment]
Support change peer.
PTAL @ngaut  @disksing

File: server/cluster.go [hunk 1]
```diff
@@ -172,6 +189,13 @@ func makeRegionSearchKey(clusterRootPath string, endKey []byte) string {
 	return strings.Join([]string{clusterRootPath, "k", encodeRegionSearchKey(endKey)}, "/")
 }
 
+func makeJobKey(clusterRootPath string, jobID uint64) string {
+	// We must guarantee the job handling order, so use %020d to format the job key,
+	// use etcd range get to get the first job and then handle it.
+	// Should we use a 8 bytes binary BigEndian instead of 20 bytes string?
+	return strings.Join([]string{clusterRootPath, "j", fmt.Sprintf("%020d", jobID)}, "/")
```
Comment (ngaut) [inline server/cluster.go]: s/j/jid

File: server/cluster_worker.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+		return errors.Errorf("post job %v fail", job)
+	}
+
+	// Tell job worker to handle the job
+	asyncNotify(c.askJobCh)
+
+	return nil
+}
+
+func (c *raftCluster) getFirstJob() (*pd_jobpd.Job, error) {
```
Comment (ngaut) [inline server/cluster_worker.go]: Remove "First"

File: server/cluster_worker.go [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	} else if !ok {
+		return nil, nil
+	}
+
+	return &job, nil
+}
+
+func (c *raftCluster) popFirstJob(job *pd_jobpd.Job) error {
```
Comment (ngaut) [inline server/cluster_worker.go]: Ditto.

File: server/cluster_worker.go [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+		}
+	}
+
+	if len(bestStores) == 0 && len(matchStores) == 0 {
+		return nil, errors.Errorf("find no store to add peer for region %v", region)
+	}
+
+	var store metapb.Store
+	// Select the store randomly, later we will do more better choice.
+	if len(bestStores) > 0 {
```
Comment (ngaut) [inline server/cluster_worker.go]: Extract to a method: chooseStore.

File: server/cluster_worker.go [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+		NodeId:  proto.Uint64(store.GetNodeId()),
+		StoreId: proto.Uint64(store.GetStoreId()),
+		PeerId:  proto.Uint64(peerID),
+	}
+
+	return peer, nil
+}
+
+// If leader is nil, we can remove any peer in the region, or else we can only remove none leader peer.
+func (c *raftCluster) handleAskChangeRemovePeer(region *metapb.Region, leader *metapb.Peer) (*metapb.Peer, error) {
```
Comment (ngaut) [inline server/cluster_worker.go]: s/handleAskChangeRemovePeer/handleRemovePeerReq/

File: server/cluster_worker.go [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+	switch req.AdminRequest.GetCmdType() {
+	case raft_cmdpb.AdminCommandType_ChangePeer:
+		return c.handleChangePeer(job)
+	default:
+		log.Errorf("invalid job command %v, ignore", req)
+		return nil
+	}
+}
+
+func (c *raftCluster) handleAskChangeAddPeer(region *metapb.Region) (*metapb.Peer, error) {
```
Comment (ngaut) [inline server/cluster_worker.go]: s/handleAskChangeAddPeer/handleAddPeerReq

File: server/command.go [hunk 7]
```diff
@@ -200,3 +200,27 @@ func (c *conn) handlePutMeta(req *pdpb.Request) (*pdpb.Response, error) {
 		PutMeta: resp,
 	}, nil
 }
+
+func (c *conn) handleAskChangePeer(req *pdpb.Request) (*pdpb.Response, error) {
+	request := req.GetAskChangePeer()
+	if request == nil {
+		return nil, errors.Errorf("invalid ask change peer command, but %v", req)
+	} else if request.Region == nil {
+		return nil, errors.Errorf("missing region for change peer")
```
Comment (ngaut) [inline server/command.go]: s/change/changing

File: server/command.go [hunk 8]
```diff
@@ -200,3 +200,27 @@ func (c *conn) handlePutMeta(req *pdpb.Request) (*pdpb.Response, error) {
...
+	} else if request.Leader == nil {
+		return nil, errors.Errorf("missing leader for change peer")
```
Comment (ngaut) [inline server/command.go]: Ditto.

File: server/conn.go [hunk 9]
```diff
@@ -116,6 +116,8 @@ func (c *conn) handleRequest(req *pdpb.Request) (*pdpb.Response, error) {
 		return c.handleGetMeta(req)
 	case pdpb.CommandType_PutMeta:
 		return c.handlePutMeta(req)
+	case pdpb.CommandType_AskChangePeer:
+		return c.handleAskChangePeer(req)
```
Comment (ngaut) [inline server/conn.go]: We may remove Ask in method name.

Comment (siddontang) [inline server/conn.go]: No, here is for Handle ask, and the worker will get the job and do the really request later.

File: server/cluster.go (hunk 1 above)
Comment (siddontang) [inline server/cluster.go]: I think j is ok, I want to use a single char for different types. And if we want to use detailed flag, I think job is better.

Comment (ngaut) [inline server/cluster.go]: Then use job please.

File: server/cluster_worker.go (hunk 4 above)
Comment (siddontang) [inline server/cluster_worker.go]: I think we should extract above codes too into chooseStore.

**Comment by siddontang** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @disksing

File: server/cluster_worker.go [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+
+	jobValue, err := proto.Marshal(job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	jobPath := makeJobKey(c.clusterRoot, jobID)
+
+	resp, err := c.s.client.Txn(context.TODO()).
+		If(c.s.leaderCmp()).
```
Comment (disksing) [inline server/cluster_worker.go]: Should we check Job ID's absence?

Comment (siddontang) [inline server/cluster_worker.go]: TODO, now we miss many checks with etcd. 
I will learn the etcd v3 deeply later and find the best way to communicate with it.

Comment (siddontang) [inline server/cluster_worker.go]: btw, here we don't check job id absent, but duplicated.

File: server/cluster_worker.go [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+		} else if existNode {
+			matchStores = append(matchStores, store)
+		} else {
+			bestStores = append(bestStores, store)
+		}
+	}
+
+	if len(bestStores) == 0 && len(matchStores) == 0 {
+		return nil, errors.Errorf("find no store to add peer for region %v", region)
+	}
```
Comment (disksing) [inline server/cluster_worker.go]: I prefer moving these lines into `chooseStore`.

Comment (siddontang) [inline server/cluster_worker.go]: no need, and the error message contains the region info, if we move to `chooseStore`, we should pass region as another arg too. 

and later we will improve the choosing strategy, so maybe later.

File: server/cluster_worker.go [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+// If leader is nil, we can remove any peer in the region, or else we can only remove none leader peer.
+func (c *raftCluster) handleRemovePeerReq(region *metapb.Region, leader *metapb.Peer) (*metapb.Peer, error) {
+	if len(region.Peers) <= 1 {
+		return nil, errors.Errorf("can not remove peer for region %v", region)
+	}
+
+	// Now we only remove the first peer, later we will do more better choice.
+	if leader == nil {
+		return region.Peers[0], nil
+	}
```
Comment (disksing) [inline server/cluster_worker.go]: `leader` will not be nil. It is already covered in `handleAskChangePeer`.

Comment (siddontang) [inline server/cluster_worker.go]: maybe later. 
But I will remove it here now.

**Comment by disksing** [comment]
LGTM

**Pull request merged**